metrics/transport.go

```diff
@@ -0,0 +1,47 @@
+package metrics
+
+import (
+	"net/http"
+	"strconv"
+	"time"
+)
+
+// Transport is an HTTP transport that would check
+// the requests and increase some metrics, cache
+// errors, etc based on the requests and responses
+type Transport struct {
+	Base    http.RoundTripper
+	metrics Provider
+}
+
+// NewTransport returns a transport using a provided http.RoundTripper as
+// the base and a metrics provider
+func NewTransport(base http.RoundTripper, metrics Provider) *Transport {
```

Comment:
One thing. When we move the `Provider` to the server package this Provider dependency is going to need the server package. Is that right or makes sense to move the transport to the server package too (hopefully not).

// @isacikgoz @agnivade

Reply:
Yeah the problem is of package boundaries. If the Transport consumes the interface, then it cannot be in the source package which provides the implementation. And if transport uses the real implementation, then the client package needs to cast it back into the main implementation. So it's problematic both ways.

Yet another way might have been to include the transport and the roundtripper interface into the main interface itself. But I don't like that.

I think this is fine.